src/main.rs

```diff
@@ -1,13 +1,186 @@
+use clap::{App, Arg};
 use std::env;
+use std::ffi::OsString;
 
 mod cli;
 
+use cli::TrinConfig;
+
+impl TrinConfig {
+    fn new() -> Self {
+        Self::new_from(std::env::args_os().into_iter()).unwrap_or_else(|e| e.exit())
+    }
+
+    fn new_from<I, T>(args: I) -> Result<Self, clap::Error>
+    where
+        I: Iterator<Item = T>,
+        T: Into<OsString> + Clone,
+    {
+        // TODO: things to configure:
+        //  - infura project id (not just env var?)
```

Comment:
Any further thoughts on this? At the moment, using just an env var makes testing slightly tricky. Some other options include a config file or a cli arg (i'm not a fan of the cli arg option, it would get rather tedious imo)

Reply:
I'm okay leaving it as env var only for now. I think we'll want support for that at some point. (One plausible use case is as a colocated endpoint for a web server, where env var would be the nicest way to configure IMO)

So I guess we'll just have to figure out a way to test where we can mock out the env var.

Reply:
Yeah, that seems reasonable. I think once we get CI setup, it'll help illuminate the best way forward. Also, from the [docs](https://doc.rust-lang.org/book/ch11-02-running-tests.html)...

> Because the tests are running at the same time, make sure your tests don’t depend on each other or on any shared state, including a shared environment, such as the current working directory or environment variables.

Which seems to me rules out any tests that check for the presence of an env variable. From my tinkering around any test that manually cleared and reset the env var (to check for presence) would cause every other test to simultaneously fail.

Reply:
Yeah, I think if we want to keep the env var long term, we would read the value in `main()` and pass it into the inner function call. Then, all layers beneath it are just tested with supplied arguments.